File: rust/src/p0938_range_sum_of_bst.rs
use std::rc::Rc;
use std::cell::RefCell;
use crate::binary_tree::TreeNode;
pub struct Solution;

impl Solution {
    pub fn range_sum_bst(root: Option<Rc<RefCell<TreeNode>>>, low: i32, high: i32) -> i32 {
        fn range_sum(root: &Option<Rc<RefCell<TreeNode>>>, low: &i32, high: &i32) -> i32 {
            match root {
                Some(root) => match &*root.borrow() {
                    TreeNode {val, left, right} if val >= low && val <= high => *val + range_sum(left, low, high) + range_sum(right, low, high),
                    TreeNode {val, right, ..} if val < low => range_sum(right, low, high),
                    TreeNode {val, left, ..} if val > high => range_sum(left, low, high),
                    _ => 0
                },
                None => 0
            }
        }
        range_sum(&root, &low, &high)
    }
}

#[cfg(test)]
mod test {
    use crate::binary_tree::TreeNode;
    use test_case::test_case;
    use super::Solution;

    #[test_case("[10,5,15,3,7,null,18]", 7, 15, 32)]
    #[test_case("[10,5,15,3,7,13,18,1,null,6]", 6, 10, 23)]
    fn solve(tree: &str, low: i32, high: i32, expected: i32) {
        let tree = TreeNode::from_str(tree);
        assert_eq!(expected, Solution::range_sum_bst(tree, low, high));
    }
}

File: dotnet/P0938_RangeSumOfBst.cs
namespace Problems;

using Problems.DataStructures;

public class P0938_RangeSumOfBst
{
    public int RangeSumBST(TreeNode? root, int low, int high) {
        return root switch {
            TreeNode {val: var val, left: var l, right: var r} when val >= low && val <= high =>
                val + RangeSumBST(l, low, high) + RangeSumBST(r, low, high),
            TreeNode {val: var val, right: var r} when val < low => RangeSumBST(r, low, high),
            TreeNode {val: var val, left: var l} when val > high => RangeSumBST(l, low, high),
            _ => 0
        };
    }

    [Theory]
    [InlineData("[10,5,15,3,7,null,18]", 7, 15, 32)]
    [InlineData("[10,5,15,3,7,13,18,1,null,6]", 6, 10, 23)]
    public void Solve(string tree, int low, int high, int expected) {
        var root = TreeNode.FromString(tree);
        Assert.Equal(expected, RangeSumBST(root, low, high));
    }    
}